pkg/monitors/jenkins/jenkins.go

```diff
@@ -0,0 +1,149 @@
+package jenkins
+
+import (
+	"context"
+	"time"
+
+	"github.com/signalfx/golib/v3/datapoint"
+	"github.com/signalfx/signalfx-agent/pkg/core/common/httpclient"
+	"github.com/signalfx/signalfx-agent/pkg/core/config"
+	"github.com/signalfx/signalfx-agent/pkg/monitors"
+	jc "github.com/signalfx/signalfx-agent/pkg/monitors/jenkins/client"
+	"github.com/signalfx/signalfx-agent/pkg/monitors/types"
+	"github.com/signalfx/signalfx-agent/pkg/utils"
+	"github.com/signalfx/signalfx-agent/pkg/utils/timeutil"
+	log "github.com/sirupsen/logrus"
+)
+
+const (
+	healthEndpoint  = "metrics/%s/healthcheck/"
+	metricsEndpoint = "metrics/%s/metrics/"
+	pingEndPoint    = "metrics/%s/ping/"
+)
+
+// Config for this monitor
+type Config struct {
+	config.MonitorConfig  `yaml:",inline" acceptsEndpoints:"true"`
+	httpclient.HTTPConfig `yaml:",inline"`
+
+	// Host of the jenkins instance
+	Host string `yaml:"host" validate:"required"`
+	// Port of the jenkins instance
+	Port string `yaml:"port" validate:"required"`
+
+	// Api key to authenticate to the jenkins metrics plugin
+	MetricsKey string `yaml:"metricsKey" validate:"required"`
+
+	// Dimension with a value that uniquely identifies the jenkins cluster
+	JenkinsCluster string `yaml:"jenkins_cluster" validate:"required"`
+	// The interval (in minutes) at which to send a dimension update (default is 5 minutes).
+	DimUpdateInterval timeutil.Duration `yaml:"dimUpdateInterval" default:"5m"`
+}
+
+type Monitor struct {
+	Output types.FilteringOutput
+	cancel context.CancelFunc
+	ctx    context.Context
+	// lastBuildSent map (key:job_name and value:build_completion_time) used to keep track of processed builds
+	lastBuildSent map[string]int64
+}
+
+var logger = utils.NewThrottledLogger(log.WithFields(log.Fields{"monitorType": monitorType}), 20*time.Second)
+
+func init() {
+	monitors.Register(&monitorMetadata, func() interface{} { return &Monitor{} }, &Config{})
+}
+
+// Configure monitor
+func (m *Monitor) Configure(conf *Config) error {
+	httpClient, err := conf.HTTPConfig.Build()
+	if err != nil {
+		return err
+	}
+	jkClient, err := jc.NewJenkinsClient(conf.Host, conf.MetricsKey, conf.Port, conf.Scheme(), httpClient)
+	if err != nil {
+		return err
+	}
+	m.ctx, m.cancel = context.WithCancel(context.Background())
+	m.lastBuildSent = make(map[string]int64)
+
+	uniqueDimeMap := map[string]string{"jenkins_cluster": conf.JenkinsCluster}
+	// anonymous function used to inject the unique dimension
+	updateAndSendDPS := func(dps ...*datapoint.Datapoint) {
+		for _, dp := range dps {
+			if dp.Dimensions == nil {
+				dp.Dimensions = uniqueDimeMap
```

Comment:
`Dime` --> `Dim`

Reply:
Also you don't want to set it to the uniqueDimeMap because logic in `Output.SendDatapoints` might modify the map and mess up the map used in this montior.  Use `utils.MergeStringMaps` which will simplify this and remove the need for `if/else`.

Reply:
done

Reply:
fixed